src/test/ui/hrtb/hrtb-cache-issue-54302.stderr

```diff
@@ -4,8 +4,8 @@ error: implementation of `Deserialize` is not general enough
 LL |     assert_deserialize_owned::<&'static str>(); //~ ERROR
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
-   = note: `&'static str` must implement `Deserialize<'0>` for any lifetime `'0`
-   = note: but `&str` only implements `Deserialize<'1>` for some lifetime `'1`
+   = note: `Deserialize<'0>` would have to be implemented for the type `&'static str`, for any lifetime `'0`
```

Comment:
I find that the reordering reads slightly worse in this case (specifically, where the trait talks about `'0` but not the self-type). I was originally thinking we would only re-order if the `'0` did not appear in the trait reference (i.e., only in the self type). But i'm not sure it's worth it to add yet more cases to this function. Thoughts?

Reply:
I can try only reordering if the `'0` appears in the self type, so that at least these existing, likely more common, error cases, don't change/regress.